File: go/medium/208_implement_trie_prefix_tree.go
// https://leetcode.com/problems/implement-trie-prefix-tree/

package medium

type Trie struct {
	root     *Trie
	children [26]*Trie
	isEnd    bool
}

func ConstructorTrie() Trie {
	return Trie{
		root:     &Trie{},
		children: [26]*Trie{},
		isEnd:    false,
	}
}

func (t *Trie) Insert(word string) {
	curr := t.root
	for _, w := range word {
		idx := w - 'a'
		if curr.children[idx] == nil {
			curr.children[idx] = &Trie{}
		}
		curr = curr.children[idx]
	}
	curr.isEnd = true
}

func (t *Trie) Search(word string) bool {
	curr := t.root
	for _, w := range word {
		idx := w - 'a'
		if curr.children[idx] == nil {
			return false
		}
		curr = curr.children[idx]
	}
	return curr.isEnd
}

func (t *Trie) StartsWith(prefix string) bool {
	curr := t.root
	for _, p := range prefix {
		idx := p - 'a'
		if curr.children[idx] == nil {
			return false
		}
		curr = curr.children[idx]
	}
	return true
}

/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */

File: java/medium/208_implement_trie_prefix_tree.java
// https://leetcode.com/problems/implement-trie-prefix-tree/

package java.medium;

class Trie {

    Node root;

    public Trie() {
        root = new Node('\0'); // dummy node
    }

    public void insert(String word) {
        Node curr = root;
        for (char x : word.toCharArray()) {
            if (curr.children[x-'a'] == null) {
                curr.children[x-'a'] = new Node(x);
            }
            curr = curr.children[x-'a'];
        }
        curr.isEnd = true;
    }

    public boolean search(String word) {
        Node res = getLast(word);
        return (res != null && res.isEnd);
    }

    public Node getLast(String word) {
        Node curr = root;
        for (char x : word.toCharArray()) {
            if (curr.children[x-'a'] == null) {
                return null;
            }
            curr = curr.children[x-'a'];
        }
        return curr;
    }

    public boolean startsWith(String prefix) {
        Node res = getLast(prefix);
        return res != null;
    }

    class Node {
        private Node[] children;
        private boolean isEnd;

        public Node(char val) {
            this.isEnd = false;
            this.children = new Node[26];
        }
    }
}